### PR TITLE
Doc-strings for some internals.

### PR DESCRIPTION
**Pull request by mschauer**
Just to get the ball rolling. Feel free to  replace or modify. Some similar doc-strings are still missing, for example for`DataType`. Also there are many undocumented exceptions, some might deserve a doc-string too, see https://gist.github.com/mschauer/f01fe4231ec0da414b7559fd05a34813

### CODE REVIEW

File: base/docs/basedocs.jl [hunk 1]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
 """
 InitError
 
+"""
+    Any::DataType
+
+`Any` is the union of all types. It has the defining property `isa(x, Any) == true` for any `x`. `Any` describes
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: Perhaps `describes therefore` -> `therefore describes`? Or `` `Any` includes the entire...``

File: base/docs/basedocs.jl [hunk 2]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+therefore the entire universe of possible values. For example `Integer` is a subset of `Any` that includes `Int`,
+`Int8`, and other concrete types.
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: Perhaps `concrete` -> `integer` to be more specific?

File: base/docs/basedocs.jl [hunk 3]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+Any
+
+"""
+    Union{}
+
+`Union{}`, the empty [`Union`](@ref) of types, is the type that has no values. That is, it has the defining
+property `isa(x, Union{}) == false` for any `x`. `Base.Bottom` is defined as its alias and the type of `Any`
+is `Core.TypeofBottom`.
+
+Examples:
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: The convention is to use `# Examples` (here and also below)

File: base/docs/basedocs.jl [hunk 4]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+    Any::DataType
+
+`Any` is the union of all types. It has the defining property `isa(x, Any) == true` for any `x`. `Any` describes
+therefore the entire universe of possible values. For example `Integer` is a subset of `Any` that includes `Int`,
+`Int8`, and other concrete types.
+"""
+Any
+
+"""
+    Union{}
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: Perhaps collapse this with the docstring of `Union{Types...}` and mention there that `Union{}` is the empty case?

File: base/docs/basedocs.jl [hunk 5]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+"""
+Union
+
+
+"""
+    UnionAll
+
+A union of types over all values of a type parameter. `UnionAll` is used to describe parametric types
+where the values of some parameters are not known.
+
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `# Examples`

File: base/docs/basedocs.jl [hunk 6]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+julia> typeof(Vector{Int})
+DataType
+```
+"""
+UnionAll
+
+"""
+    ::
+With the `::`-operator type annotations are attached to expressions and variables in programs.
+See manual/types/#type-declarations.
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `See the manual section on [Type Declarations](@ref)`

File: base/docs/basedocs.jl [hunk 7]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+
+Outside of declarations `::` is used to assert that expressions and variables in programs have a given type.
+
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `# Examples`

File: base/docs/basedocs.jl [hunk 8]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+julia> (1+2)::Int
+3
+```
+"""
+kw"::"
+
+"""
+    Vararg{T,N}
+
+The last parameter of a tuple type [`Tuple`](@ref) can be the special type `Vararg`. `Vararg` which denotes any
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: Second sentence here seems a bit weird?

File: base/docs/basedocs.jl [hunk 9]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+number of trailing elements. The type `Vararg{T,N}` corresponds to exactly `N` elements of type `T`.
+`Vararg{T}` correspondsvto zero or more elements of type `T`. `Vararg` tuple types are used to represent the
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `correspondsvto` -> `corresponds to`

File: base/docs/basedocs.jl [hunk 10]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+arguments accepted byvvarargs methods (see vararg-functions in the manual).
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `byvvarargs` -> `by vararg`

Comment (fredrikekre) [inline base/docs/basedocs.jl]: `(see the section on [Varargs Functions](@ref) in the manual.)`

File: base/docs/basedocs.jl [hunk 11]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `# Examples`

File: base/docs/basedocs.jl [hunk 12]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+Tuples are an abstraction of the arguments of a function – without the function itself. The salient aspects of
+a function's arguments are their order and their types. Therefore a tuple type is similar to a parameterized
+immutable type where each parameter is the type of one field. Tuple types may have any number of parameters.
+
+Tuple types are covariant in their parameters: `Tuple{Int}` is a subtype of `Tuple{Any}`. Therefore `Tuple{Any}`
+is considered an abstract type, and tuple types are only concrete if their parameters are. Tuples do not have
+field names; fields are only accessed by index.
+
+See manual, section "Tuple Types".
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: `See manual, section [Tuple Types](@ref).`

**Review by fredrikekre — COMMENTED** [review]
Great! :+1: 
Since these are all new docstrings they need to be added on appropriate location in the stdlib doc index, see https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#adding-a-new-docstring-to-base

File: base/docs/basedocs.jl [hunk 13]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+therefore the entire universe of possible values. For example `Integer` is a subset of `Any` that includes `Int`,
+`Int8`, and other concrete types.
+"""
+Any
+
+"""
+    Union{}
+
+`Union{}`, the empty [`Union`](@ref) of types, is the type that has no values. That is, it has the defining
+property `isa(x, Union{}) == false` for any `x`. `Base.Bottom` is defined as its alias and the type of `Any`
```
Comment (tkelman) [inline base/docs/basedocs.jl]: type of `Union{}` or `Base.Bottom`, not `Any`, right?

File: base/docs/basedocs.jl [hunk 14]
```diff
@@ -1261,4 +1261,126 @@ available in the `.error` field.
...
+
+julia> typeof(Vector{Int})
+DataType
+```
+"""
+UnionAll
+
+"""
+    ::
+With the `::`-operator type annotations are attached to expressions and variables in programs.
```
Comment (tkelman) [inline base/docs/basedocs.jl]: blank line between the operator and start of description

File: base/docs/basedocs.jl (hunk 4 above)
Comment (mschauer) [inline base/docs/basedocs.jl]: I think Union{} is a bit special as it has its own type.

**Comment by mschauer** [comment]
@fredrikekre Thank you!

> Since these are all new docstrings they need to be added on appropriate location in the stdlib doc index, see https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#adding-a-new-docstring-to-base

I am not sure, aren't these things already better treated in the manual section? I at least would not look for them in the library section. Similarly the section on "punctation" which mixes some operators (but not the mathematical one) and syntactical elements in the library section seems odd to me...

**Comment by mschauer** [comment]
Is it correct that a function which does not return, has return type `Union{}` whereas a function that returns but returns nothing (= `nothing`) has return type `Void`?

File: base/docs/basedocs.jl [hunk 15]
```diff
@@ -1288,9 +1288,10 @@ kw"Union{}"
 """
     Union{Types...}
 
-A type union is an abstract type which includes all instances of any of its argument types.
+A type union is an abstract type which includes all instances of any of its argument types. The empty
+union [`Union{}`](@ref) is the bottom type of julia.
```
Comment (tkelman) [inline base/docs/basedocs.jl]: would capitalize this as it refers to the language properties, not the command-line executable

File: base/docs/basedocs.jl (hunk 4 above)
Comment (mschauer) [inline base/docs/basedocs.jl]: In addition this docstring is now attached to `Bottom` as well.

**Comment by fredrikekre** [comment]
> I am not sure, aren't these things already better treated in the manual section? I at least would not look for them in the library section. Similarly the section on "punctation" which mixes some operators (but not the mathematical one) and syntactical elements in the library section seems odd to me...

It is nice to be able to find these when searching the documentation, so should definitely be added. I'm not sure where though, looks like some of these could fit in `stdlib/base.md`?

**Comment by JeffBezanson** [comment]
Looks fine to me, ready to merge?

**Review by fredrikekre — APPROVED** [review]

**Comment by mschauer** [comment]
`make` reports an error but I did not yet figure out what is going on: 
```
Building HTML documentation.
/cygdrive/c/projects/julia/usr/bin/julia.exe `cygpath -w /cygdrive/c/projects/julia/doc/make.jl` > docbuild.log 2>&1
make[2]: *** [Makefile:34: html] Error 1
make[2]: Leaving directory '/cygdrive/c/projects/julia/doc'
make[1]: *** [Makefile:119: docs] Error 2
make[1]: Leaving directory '/cygdrive/c/projects/julia'
```

**Comment by fredrikekre** [comment]
Seems like this is the issue:
```
 !! Duplicate docs found for 'Core.Union{}'. [src/stdlib/base.md]
 !! No docs found for 'Core.Tuple'. [src/stdlib/base.md]
```

**Comment by mschauer** [comment]
Ah, these are keywords.

File: base/docs/basedocs.jl [hunk 16]
```diff
@@ -1261,4 +1261,131 @@ available in the `.error` field.
...
+a function's arguments are their order and their types. Therefore a tuple type is similar to a parameterized
+immutable type where each parameter is the type of one field. Tuple types may have any number of parameters.
+
+Tuple types are covariant in their parameters: `Tuple{Int}` is a subtype of `Tuple{Any}`. Therefore `Tuple{Any}`
+is considered an abstract type, and tuple types are only concrete if their parameters are. Tuples do not have
+field names; fields are only accessed by index.
+
+See the manual section on [Tuple Types](@ref).
+"""
+kw"Tuple"
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: This can be `Core.Tuple`

Comment (mschauer) [inline base/docs/basedocs.jl]: Yes, I changed it already, amending in a moment.

**Comment by fredrikekre** [comment]
Heh, if you want you could apply this
```diff
diff --git a/test/docs.jl b/test/docs.jl
index 635aa36..395a03b 100644
--- a/test/docs.jl
+++ b/test/docs.jl
@@ -1055,3 +1055,6 @@ end
 end
 @test Main.f23011() == 2
 @test docstrings_equal(@doc(Main.f23011), doc"second")
+
+# test that retrieving docs for Base.Bottom does not crash julia
+@test contains(sprint(show, @doc(Base.Bottom)), "Union{}")
```

Because this PR actually fixes the following problem I was about to file an issue for:
```
help?> Base.Bottom
Unreachable reached at 0x7fe7e8b520e4

signal (4): Illegal instruction
while loading no file, in expression starting on line 212
summarize at ./docs/Docs.jl:431
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
summarize at ./docs/Docs.jl:364
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
doc at ./docs/Docs.jl:304
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_invoke at /home/fredrik/julia-master/src/gf.c:41
doc at ./docs/Docs.jl:282
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
do_call at /home/fredrik/julia-master/src/interpreter.c:70
eval at /home/fredrik/julia-master/src/interpreter.c:262
eval_body at /home/fredrik/julia-master/src/interpreter.c:543
jl_interpret_toplevel_thunk at /home/fredrik/julia-master/src/interpreter.c:689
jl_toplevel_eval_flex at /home/fredrik/julia-master/src/toplevel.c:622
jl_toplevel_eval_in at /home/fredrik/julia-master/src/builtins.c:510
eval at ./repl/REPL.jl:3
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
eval_user_input at ./repl/REPL.jl:69
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
macro expansion at ./repl/REPL.jl:100 [inlined]
#1 at ./event.jl:96
jl_call_fptr_internal at /home/fredrik/julia-master/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/fredrik/julia-master/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/fredrik/julia-master/src/gf.c:1940
jl_apply at /home/fredrik/julia-master/src/julia.h:1447 [inlined]
start_task at /home/fredrik/julia-master/src/task.c:268
unknown function (ip: 0xffffffffffffffff)
Allocations: 2771563 (Pool: 2771052; Big: 511); GC: 5
Illegal instruction (core dumped)
```

**Comment by mschauer** [comment]
Isn't that a deeper bug? 

```
 julia> Docs.summarize(STDERR, Union{}, Base.Bottom)
 `Union{}` is of type `Core.TypeofBottom`.
  
# Summary
struct Core.TypeofBottom <: Type{Union{}}

# Supertype Hierarchy
Core.TypeofBottom <: Type{Union{}} <: Any
 
Unreachable reached at 0x118d63d98

signal (4): Illegal instruction: 4
while loading no file, in expression starting on line 0
summarize at ./docs/Docs.jl:431
do_call at /Users/schauermr/julia/src/interpreter.c:70
eval at /Users/schauermr/julia/src/interpreter.c:262
jl_interpret_toplevel_expr_in at /Users/schauermr/julia/src/interpreter.c:50
jl_toplevel_eval_flex at /Users/schauermr/julia/src/toplevel.c:607
jl_toplevel_eval_in at /Users/schauermr/julia/src/builtins.c:505
eval at ./repl/REPL.jl:3
eval_user_input at ./repl/REPL.jl:69
macro expansion at ./repl/REPL.jl:100 [inlined]
#1 at ./event.jl:96
jl_apply at /Users/schauermr/julia/src/./julia.h:1447 [inlined]
start_task at /Users/schauermr/julia/src/task.c:268
Allocations: 1113089 (Pool: 1112916; Big: 173); GC: 1
Illegal instruction: 4
```

**Comment by fredrikekre** [comment]
Probably, I tried to step through the code, but it did not crash when doing it one step at a time...

**Comment by mschauer** [comment]
> Probably, I tried to step through the code, but it did not crash when doing it one step at a time...

Same here.

**Comment by mschauer** [comment]
Conclusion: Let's leave this pr as it is and file an issue for the crash.

**Pull request merged**